cluster-autoscaler/cloudprovider/gce/autoscaling_gce_client.go

```
@@ -753,14 +753,19 @@ func (client *autoscalingGceClientV1) FetchAvailableCpuPlatforms() (map[string][
 	return availableCpuPlatforms, nil
 }
 
-func (client *autoscalingGceClientV1) FetchAvailableDiskTypes() (map[string][]string, error) {
+// FetchAvailableDiskTypes returns a map of zone to available disk types filtered by region.
+func (client *autoscalingGceClientV1) FetchAvailableDiskTypes(region string) (map[string][]string, error) {
 	availableDiskTypes := make(map[string][]string)
 
-	req := client.gceService.DiskTypes.AggregatedList(client.projectId)
+	// We use a filter to retrieve disk types available within a specific region.
+	// This filter leverages the GCE naming convention where zone names are prefixed with their region (e.g., "us-central1-a" is in the "us-central1" region).
+	// By using the provided region, the filter effectively selects all matching zones within that region.
+	// For example, if region is "us-central1", the filter will match all zones in that region (e.g., "us-central1-a", "us-central1-b").
+	filter := fmt.Sprintf("zone eq .+%s.*", region)
+	req := client.gceService.DiskTypes.AggregatedList(client.projectId).Filter(filter).ReturnPartialSuccess(true)
```

How do we distinguish partial failure with information from `region` missing from a full success?

---

TBH I don't know exactly. When zone/region is filtered out we get a warning with code "NO_RESULTS_ON_PAGE". I guess in case of failure we will see another error code but need to verify it. But why we need this information? To log some sort of error in case of outages?

---

But based on documentation we cannot get not full list of diskTypes for particular zone, for example. It says "when partial success behavior is enabled, aggregatedList for a single zone scope either returns all resources in the zone or no resources, with an error code." So if we didn't get warning code for the zone it means we got full list of disktypes.

---

We already have this check https://github.com/kubernetes/autoscaler/pull/7923/files#diff-03130b1f2758bd6e3f5a47986ebc8a2fbb5df65a58ece878afd2efa16b8850c5R769. If Zone field is empty it means smth went wrong for this particular zone. And we do not return any data for this zone.

Later in GKE code we fetching and saving in cache data from diskTypes API. If there is no data for requested zone in cache we will send request to API again.

---

Yes, let's be clear about the error returned. If one is interested in disks from a given region, the function fetching them should return an error upon failure to get a complete response for that region.